pkg/intermediate/aggregate.go

```diff
@@ -124,74 +140,272 @@ func (a *AggregationProcess) DeleteFlowKeyFromMap(flowKey FlowKey) {
+				}
+			}
+		}
+	}
+}
+
+func (a *AggregationProcess) addFieldsForStatsAggregation(record entities.Record, fillSrcStats, fillDstStats bool) error {
+	if a.aggregateElements == nil {
+		return nil
+	}
+	statsElementList := a.aggregateElements.statsElements
+	antreaSourceStatsElements := a.aggregateElements.aggregatedSourceStatsElements
+	antreaDestinationStatsElements := a.aggregateElements.aggregatedDestinationStatsElements
+	antreaElements := append(antreaSourceStatsElements, antreaDestinationStatsElements...)
+
+	for _, element := range antreaElements {
+		// Get the new info element from Antrea registry.
+		// TODO: Take antrea registry enterpriseID as input to make this generic.
+		ie, err := registry.GetInfoElement(element, registry.AntreaEnterpriseID)
+		if err != nil {
+			return err
+		}
+		value := new(bytes.Buffer)
+		if err = util.Encode(value, binary.BigEndian, uint64(0)); err != nil {
+			return err
+		}
+		ieWithValue := entities.NewInfoElementWithValue(ie, value)
+		_, err = record.AddInfoElement(ieWithValue, true)
+		if err != nil {
+			return err
+		}
+	}
+	// Initialize the values of newly added stats info elements.
+	for i, element := range statsElementList {
+		if ieWithValue, exist := record.GetInfoElementWithValue(element); exist {
+			// Initialize the corresponding source element in antreaStatsElement list.
+			if fillSrcStats {
+				existingIeWithValue, _ := record.GetInfoElementWithValue(antreaSourceStatsElements[i])
+				existingIeWithValue.Value = ieWithValue.Value
+			}
+			// Initialize the corresponding destination element in antreaStatsElement list.
+			if fillDstStats {
+				existingIeWithValue, _ := record.GetInfoElementWithValue(antreaDestinationStatsElements[i])
+				existingIeWithValue.Value = ieWithValue.Value
+			}
+		}
 	}
+	return nil
 }
 
-// addRecordToMap is currently used only in correlateRecords().
-// For other uses, please acquire the flowKeyRecordLock for protection.
-func (a *AggregationProcess) addRecordToMap(flowKey FlowKey, record entities.Record) {
-	if _, exist := a.flowKeyRecordMap[flowKey]; !exist {
-		a.flowKeyRecordMap[flowKey] = make([]entities.Record, 0)
+
+// isRecordIntraNode returns true if record belongs to intra-node flow.
+func isRecordIntraNode(record entities.Record) bool {
+	srcIEWithValue, exist := record.GetInfoElementWithValue("sourcePodName")
+	if exist && srcIEWithValue.Value != "" {
+		dstIEWithValue, exist := record.GetInfoElementWithValue("destinationPodName")
+		if exist && dstIEWithValue.Value != "" {
+			return true
+		}
 	}
-	a.flowKeyRecordMap[flowKey] = append(a.flowKeyRecordMap[flowKey], record)
+	return false
 }
 
+// isRecordFromSrc returns true if record belongs to inter-node flow and from source node.
 func isRecordFromSrc(record entities.Record) bool {
+	if isRecordIntraNode(record) {
+		return false
+	}
+	ieWithValue, exist := record.GetInfoElementWithValue("destinationPodName")
+	if exist && ieWithValue.Value == "" {
```

Comment:
Should this be
    `if !exist || ieWithValue.Value == "" {`
?

Reply:
We expect destinationPodName to be present. Again we presume same template use case on both source and destination nodes. 
In future, we have to extend where there are different templates for source and destination node, but we probably have to consider different cases. We need to have one template intra-node flow record, one template for inter-node source node flow record, one template for inter-node destination flow record etc. This case is not supported.